### PR TITLE
Sorting by position on product search via REST API does not work.

### LINKED ISSUE #31591: Sorting by position on product search via REST API does not work
### Preconditions
1. Magento 2.4-develop
2. Magento 2.3.3
3. Vanilla Install

### Steps to reproduce
1. Allow Anonymous Guest Access
From : Stores -> Configuration -> Services -> Magento Web API -> Web API Security

2. Assign position value to products 
from : Catalog -> Categories -> Select Category -> Products in Category

3. Hit this endpoint:
`GET : <domain name>/rest/V1/products/?searchCriteria[filterGroups][0][filters][0][field]= category_id&searchCriteria[filterGroups][0][filters][0][value]=3&searchCriteria[sortOrders][0][field]= position&searchCriteria[sortOrders][0][direction]=ASC&searchCriteria[pageSize]=5`

Screenshot from 2021-01-08 10-44-33

### Expected result :heavy_check_mark: 
The product's list should be sorted by "position" parameter

### Actual result: :heavy_multiplication_x: 
The product's list is not sorted by the "position" parameter

Search Criteria, as returned in the response 
```
"search_criteria": {
    "filter_groups": [
      {
        "filters": [
          {
            "field": "category_id",
            "value": "3",
            "condition_type": "eq"
          }
        ]
      }
    ],
    "sort_orders": [
      {
        "field": "position",
        "direction": "ASC"
      }
    ],
    "page_size": 5
  }
```

---

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [x] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by engcom-Kilo**
### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/31591

### Manual testing scenarios (*)

1. see https://github.com/magento/magento2/issues/31591

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by danielrussob** [comment]
@magento give me test instance

**Comment by danielrussob** [comment]
@magento give me 2.4-develop instance

**Comment by danielrussob** [comment]
Issue reproduced on vanilla 2.4

**Comment by danielrussob** [comment]
Solved on new instance

File: app/code/Magento/Catalog/Model/ProductRepository.php [hunk 1]
```diff
@@ -856,4 +858,36 @@ private function saveProduct($product): void
...
+     * @param Collection $collection
+     * @param SearchCriteriaInterface $searchCriteria
+     * @return void
+     */
+    private function joinPositionField(
+        Collection $collection,
+        SearchCriteriaInterface $searchCriteria
+    ): void {
+        $categoryIds = [[]];
+        foreach ($searchCriteria->getFilterGroups() as $filterGroup) {
```
Comment (danielrussob) [inline app/code/Magento/Catalog/Model/ProductRepository.php:874]: Warning: Invalid argument supplied for foreach() in /var/www/html/app/code/Magento/Catalog/Model/ProductRepository.php:874 on test https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/31629/1987de3af131eea1fc33bf4caa363d0b/Unit/allure-report-ce/index.html#categories/50ea141dbb5485cec1df0a24a3a2dc38/6184ee0e02b630f2/

you could wrap with:

if ($searchCriteria->getFilterGroups()) {
... foreach ...
}

**Review by danielrussob — CHANGES_REQUESTED** [review]

**Comment by aleron75** [comment]
The failing unit test has to be fixed. 
Other failing tests don't seem to be related to the changes in this PR instead.

File: app/code/Magento/Catalog/Model/ProductRepository.php (hunk 1 above)
Comment (engcom-Kilo) [reply app/code/Magento/Catalog/Model/ProductRepository.php:874]: @danielrussob 
Thanks for review. There is no need to add such verification, $searchCriteria->getFilterGroups() returns array in any case. 
https://github.com/magento/magento2/blob/63434d2997f9625238d283d5dcaebfe918020546/lib/internal/Magento/Framework/Api/Search/FilterGroup.php#L29
Problem with unit test

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Review by zakdma — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

1. Allow Anonymous Guest Access
From : Stores -> Configuration -> Services -> Magento Web API -> Web API Security

2. Assign position value to products 
from : Catalog -> Categories -> Select Category -> Products in Category




3. Hit this endpoint:
`GET : <domain name>/rest/V1/products/?searchCriteria[filterGroups][0][filters][0][field]= category_id&searchCriteria[filterGroups][0][filters][0][value]=3&searchCriteria[sortOrders][0][field]= position&searchCriteria[sortOrders][0][direction]=ASC&searchCriteria[pageSize]=5`

Screenshot from 2021-01-08 10-44-33

### Before: :heavy_multiplication_x: 
The product's list is not sorted by the "position" parameter


### After: :heavy_check_mark: 
The product's list is sorted by the "position" parameter correctly

**_ASC_**


**_DESC_**


Sorting by name, price ...  work as previous (expected).

**Review by aleron75 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @aleron75, thank you for the review.
[ENGCOM-8629](https://jira.corp.magento.com/browse/ENGCOM-8629) has been created to process this Pull Request

**Review by danielrussob — APPROVED** [review]

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-31629`.

**Pull request merged**